service/matching/task_validation.go

```diff
@@ -0,0 +1,195 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+
+	"go.temporal.io/server/api/historyservice/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/primitives/timestamp"
+)
+
+const (
+	taskReaderOfferTimeout        = 60 * time.Second
+	taskReaderValidationThreshold = 600 * time.Second
+)
+
+type (
+	taskValidator interface {
+		maybeValidate(
+			task *persistencespb.AllocatedTaskInfo,
+			taskType enumspb.TaskQueueType,
+		) bool
+	}
+
+	taskValidationInfo struct {
+		taskID         int64
+		validationTime time.Time
+	}
+
+	taskValidatorImpl struct {
+		historyClient historyservice.HistoryServiceClient
+
+		lastValidatedTaskInfo taskValidationInfo
+	}
+)
+
+func newTaskValidator(
+	historyClient historyservice.HistoryServiceClient,
+) *taskValidatorImpl {
+	return &taskValidatorImpl{
+		historyClient: historyClient,
+	}
+}
+
+// check if a task has expired / is valid
+// if return false, then task is invalid and should be discarded
+// if return true, then task is *not invalid*, and should be dispatched
+//
+// a task is invalid if this task is already failed; timeout; completed, etc
+// a task is *not invalid* if this task can be started, or caller cannot verify the validity
+func (v *taskValidatorImpl) maybeValidate(
+	task *persistencespb.AllocatedTaskInfo,
+	taskType enumspb.TaskQueueType,
+) bool {
+	if IsTaskExpired(task) {
+		return false
+	}
+	if !v.preValidate(task) {
+		return true
+	}
+	valid, err := v.isTaskValid(task, taskType)
+	if err != nil {
+		return true
+	}
+	v.postValidate(task)
+	return valid
+}
+
+// preValidate track a task and return if validation should be done
+func (v *taskValidatorImpl) preValidate(
+	task *persistencespb.AllocatedTaskInfo,
+) bool {
+	if v.lastValidatedTaskInfo.taskID != task.TaskId {
+		// first time seen the task, caller should try to dispatch first
+		if task.Data.CreateTime != nil {
+			v.lastValidatedTaskInfo = taskValidationInfo{
+				taskID:         task.TaskId,
+				validationTime: *task.Data.CreateTime, // task is valid when created
+			}
+		} else {
+			v.lastValidatedTaskInfo = taskValidationInfo{
+				taskID:         task.TaskId,
+				validationTime: time.Now().UTC(), // if no creation time specified, use now
+			}
+		}
+		return false
+	}
+
+	// this task has been validated before
+	if time.Since(v.lastValidatedTaskInfo.validationTime) > taskReaderValidationThreshold {
+		return true
+	}
+	return false
```

Comment:
could simplify.. but I understand if you want to leave it open to more conditions later
```suggestion
	return time.Since(v.lastValidatedTaskInfo.validationTime) > taskReaderValidationThreshold
```